flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerHeap.java

```diff
@@ -0,0 +1,504 @@
+	 */
+	private final int totalNumberOfKeyGroups;
+
+
+	/**
+	 * Creates an empty {@link InternalTimerHeap} with the requested initial capacity.
+	 *
+	 * @param minimumCapacity the minimum and initial capacity of this priority queue.
+	 */
+	@SuppressWarnings("unchecked")
+	InternalTimerHeap(
+		@Nonnegative int minimumCapacity,
+		@Nonnull KeyGroupRange keyGroupRange,
+		@Nonnegative int totalNumberOfKeyGroups) {
+
+		this.totalNumberOfKeyGroups = totalNumberOfKeyGroups;
+		this.keyGroupRange = keyGroupRange;
+
+		final int keyGroupsInLocalRange = keyGroupRange.getNumberOfKeyGroups();
+		final int deduplicationSetSize = 1 + minimumCapacity / keyGroupsInLocalRange;
+		this.deduplicationMapsByKeyGroup = new HashMap[keyGroupsInLocalRange];
+		for (int i = 0; i < keyGroupsInLocalRange; ++i) {
+			deduplicationMapsByKeyGroup[i] = new HashMap<>(deduplicationSetSize);
+		}
+
+		this.queue = new TimerHeapInternalTimer[1 + minimumCapacity];
+	}
+
+	/**
+	 * @see Set#add(Object)
+	 */
+	@Override
+	public boolean add(@Nonnull TimerHeapInternalTimer<K, N> timer) {
+
+		if (getDedupMapForKeyGroup(timer).putIfAbsent(timer, timer) == null) {
+			final int newSize = ++this.size;
+			checkCapacity(newSize);
+			moveElementToIdx(timer, newSize);
+			siftUp(newSize);
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	/**
+	 * This behaves like {@link #add(TimerHeapInternalTimer)}.
+	 */
+	@Override
+	public boolean offer(@Nonnull TimerHeapInternalTimer<K, N> k) {
+		return add(k);
+	}
+
+	@Nullable
+	@Override
+	public TimerHeapInternalTimer<K, N> poll() {
+		return size() > 0 ? removeElementAtIndex(1) : null;
+	}
+
+	@Nonnull
+	@Override
+	public TimerHeapInternalTimer<K, N> remove() {
+		TimerHeapInternalTimer<K, N> pollResult = poll();
+		if (pollResult != null) {
+			return pollResult;
+		} else {
+			throw new NoSuchElementException("InternalTimerPriorityQueue is empty.");
+		}
+	}
+
+	@Nullable
+	@Override
+	public TimerHeapInternalTimer<K, N> peek() {
+		return size() > 0 ? queue[1] : null;
+	}
+
+	@Nonnull
+	@Override
+	public TimerHeapInternalTimer<K, N> element() {
+		TimerHeapInternalTimer<K, N> peekResult = peek();
+		if (peekResult != null) {
+			return peekResult;
+		} else {
+			throw new NoSuchElementException("InternalTimerPriorityQueue is empty.");
+		}
+	}
+
+	@Override
+	public boolean isEmpty() {
+		return size() == 0;
+	}
+
+	@Override
+	public boolean contains(@Nullable Object o) {
+		return (o instanceof TimerHeapInternalTimer)
+			&& getDedupMapForKeyGroup((TimerHeapInternalTimer<?, ?>) o).containsKey(o);
+	}
+
+	@Override
+	public boolean remove(@Nullable Object o) {
+		if (o instanceof TimerHeapInternalTimer) {
+			return removeInternal((TimerHeapInternalTimer<?, ?>) o);
+		}
+		return false;
+	}
+
+	@Override
+	public boolean addAll(@Nullable Collection<? extends TimerHeapInternalTimer<K, N>> timers) {
+
+		if (timers == null) {
+			return true;
+		}
+
+		if (timers.size() > queue.length) {
```

Comment:
Good question, what the right choice here is because it might also be that all elements in timers are duplicates and already contained. Not sure what the optimal choice is here.

Reply:
I think maybe `if (time.size() + size() > queue.length)` is better, because we can ensure that we only need to rescale the array once with this approach. With the current version, we may need to rescale it multi times, what do you think?

Reply:
Yes, the downside is, if all elements are already contained we waste the space. Anyways, in practise this method is currently only used to bulk load empty heaps, so we can also go for your suggestion.